### PR TITLE
ec2_vpc_nat_gateway: Add tags management

### LINKED ISSUE #44339: Update "ec2_vpc_nat_gateway" module to support AWS tags
##### SUMMARY
Update the "ec2_vpc_nat_gateway" module to support AWS tags like the other AWS modules including, to name a few, the "ec2_vpc_net" module, "ec2_vpc_igw" module, and "ec2_vpc_subnet" module.  We have an internal requirement within our company to tag everything and I am not seeing that option for Ansible created NAT Gateways, even though the functionality is provided on every other AWS module I have used so far.

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
ec2_vpc_nat_gateway

##### ANSIBLE VERSION
```
ansible 2.6.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, May 31 2018, 09:41:32) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### CONFIGURATION
None

##### OS / ENVIRONMENT
Red Hat Enterprise Linux Server release 7.5 (Maipo)

##### STEPS TO REPRODUCE
Attempt to run the following task using the "ec2_vpc_nat_gateway" module with included "tags" key/value pairs.

```yaml
- name: Create NAT Gateway, allocate new EIP, in VPC
  ec2_vpc_nat_gateway:
    state: present
    # NAT Gateways being deployed in DMZ subnets
    subnet_id: "{{ subnet_dmz_id[item.subnet_name] }}"
    region: "{{ vpc_region }}"
    wait: yes
    if_exist_do_not_create: true
    tags:
      Name: "{{ item.name }}"
      Role: "{{ role_tag }}"
      Team: "{{ team_tag }}"
      Product Area: "{{ product_area_tag }}"
      Portfolio: "{{ portfolio_tag }}"
      Created By: "{{ created_by }}"
  with_items: "{{ nat_gateway }}"
  register: new_nat_gateway
```

##### EXPECTED RESULTS
AWS NAT Gateways are created and their tags are updated with the provided "tags" key/value pairs included under the "ec2_vpc_nat_gateway" module.

##### ACTUAL RESULTS
The "ec2_vpc_nat_gateway" module does not support the "tags" parameters as shown below.
```
{
    "_ansible_parsed": true,
    "changed": false,
    "_ansible_no_log": false,
    "_ansible_item_result": true,
    "item": {
        "subnet_name": "DMZ_Subnet_A",
        "name": "NAT_Gateway_A"
    },
    "_ansible_item_label": {
        "subnet_name": "DMZ_Subnet_A",
        "name": "NAT_Gateway_A"
    },
    "invocation": {
        "module_args": {
            "tags": {
                "Name": "NAT_Gateway_A",
                "Role": "splunk-proof-of-concept",
                "Created By": "ansible",
                "Team": "Engineering Tools",
                "Product Area": "Engineering Tools",
                "Portfolio": "Shared Platform and Operations"
            },
            "subnet_id": "subnet-XXXX",
            "region": "us-east-1",
            "if_exist_do_not_create": true,
            "state": "present",
            "wait": true
        }
    },
    "_ansible_ignore_errors": null,
    "msg": "Unsupported parameters for (ec2_vpc_nat_gateway) module: tags Supported parameters include: allocation_id, aws_access_key, aws_secret_key, client_token, ec2_url, eip_address, if_exist_do_not_create, nat_gateway_id, profile, region, release_eip, security_token, state, subnet_id, validate_certs, wait, wait_timeout"
}
```

### PR DESCRIPTION
**Pull request by ZuZuD**
##### SUMMARY
Update the "ec2_vpc_nat_gateway" module to support AWS tags like the other AWS modules including, to name a few, the "ec2_vpc_net" module, "ec2_vpc_igw" module, and "ec2_vpc_subnet" module. 

Fixes #44339

##### COMPONENT NAME
ec2_vpc_nat_gateway

##### ADDITIONAL INFORMATION
Add the tags dictionary and the purge_tags boolean arguments.

### CODE REVIEW

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/87244/69/tests

**Comment by ZuZuD** [comment]
> CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/87244/69/tests

I will re-write the tests for this module.

**Comment by ibaldonl** [comment]
Whats missing for this to be merged? Thanks!!!

**Comment by MQasimSarfraz** [comment]
Would be great if we merge it! :+1:

**Comment by stevenscg** [comment]
About to launch some new gateways for our environment and this would really help avoid operator intervention.

**Comment by healem** [comment]
This PR has been ready to go for months, but hasn't been merged.  What needs to happen to move this along?

File: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py [hunk 1]
```diff
@@ -66,6 +66,23 @@
       - Optional unique token to be used during create to ensure idempotency.
         When specifying this option, ensure you specify the eip_address parameter
         as well otherwise any subsequent runs will fail.
+  tags:
+    description:
+      - >
+        A dictionary of resource tags of the form: { tag1_name: 'tag1_value', tag2_name: 'tag2_value' }
+    required: false
+    aliases: [ "resource_tags" ]
+    version_added: "2.8"
```
Comment (healem) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py]: Does this have to be 2.9, now that 2.8 is entering release?

File: test/integration/targets/ec2_vpc_nat_gateway/tasks/main.yml [hunk 2]
```diff
@@ -1,82 +1,217 @@
-# The tests for this module are incomplete.
-# The tests below were migrated from unit tests.
-# They take advantage of hard-coded results within the module to trigger both changed and unchanged responses.
-# They were migrated to maintain test coverage while removing unit tests that depended on use of TaskQueueManager.
+---
```
Comment (healem) [inline test/integration/targets/ec2_vpc_nat_gateway/tasks/main.yml:1]: Nice job on the integration tests.

File: test/integration/targets/ec2_vpc_nat_gateway/tasks/main.yml [hunk 3]
```diff
@@ -1,82 +1,217 @@
...
+  
+  - name: assert tag is added
+    assert:
+      that:
+          - 'vpc_nat_gateway_tagged.changed == True'
+  
+  # ============================================================
+  - name: Pause to collect updated facts
+    pause: 
+      seconds: 3
```
Comment (healem) [inline test/integration/targets/ec2_vpc_nat_gateway/tasks/main.yml]: In my experience, pauses in tests creates unstable results - sometimes they pass, sometimes not.  Is the pause really needed?  If so, can you do something more deterministic, like put a waiter on the previous task?

**Comment by Maelstromeous** [comment]
This is a big gap in consistency in Ansible's AWS tagging. Please merge this!

File: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py [hunk 4]
```diff
@@ -925,6 +946,59 @@ def remove(client, nat_gateway_id, wait=False, wait_timeout=0,
...
+    result = response['Tags']
+    if result:
+        result = boto3_tag_list_to_ansible_dict(result)
+    else:
+        result = {}
+    return result
+
+
 def main():
     argument_spec = ec2_argument_spec()
```
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:1003]: ec2_argument_spec is no longer needed after moving to AnsibleAWSModule

File: test/integration/targets/ec2_vpc_nat_gateway/tasks/main.yml [hunk 5]
```diff
@@ -1,82 +1,212 @@
...
+- block:
+  - name: set up aws connection info
+    set_fact:
+      aws_connection_info: &aws_connection_info
+        aws_access_key: "{{ aws_access_key }}"
+        aws_secret_key: "{{ aws_secret_key }}"
+        security_token: "{{ security_token }}"
+        region: "{{ aws_region }}"
+    no_log: yes
+
```
Comment (tremble) [inline test/integration/targets/ec2_vpc_nat_gateway/tasks/main.yml:11]: This is better dome using module_defaults (please also make sure you use `security_token: "{{ security_token | default(omit) }}"`)

File: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py [hunk 6]
```diff
@@ -966,6 +1038,8 @@ def main():
     release_eip = module.params.get('release_eip')
     client_token = module.params.get('client_token')
     if_exist_do_not_create = module.params.get('if_exist_do_not_create')
+    purge_tags = module.params.get('purge_tags')
+    tags = module.params.get('tags')
 
     try:
         region, ec2_url, aws_connect_kwargs = (
```
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:1045]: The contents of this try block can be replaced by `module.client('ec2')`

File: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py [hunk 7]
```diff
@@ -183,28 +192,40 @@
...
+import traceback
 
 try:
     import botocore
 except ImportError:
-    pass  # caught by imported HAS_BOTO3
+    pass  # caught by imported ec2.HAS_BOTO3
 
+from ansible.module_utils._text import to_native
 from ansible.module_utils.basic import AnsibleModule
```
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:223]: No longer needed

File: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py [hunk 8]
```diff
@@ -183,28 +192,40 @@
...
+from ansible.module_utils.aws.core import AnsibleAWSModule
 from ansible.module_utils.ec2 import (ec2_argument_spec, get_aws_connection_info, boto3_conn,
```
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:225]: Once you replace the connection block below with module.client('ec2') you don't need these 3 imports...

File: test/integration/targets/ec2_vpc_nat_gateway/tasks/main.yml [hunk 9]
```diff
@@ -1,82 +1,212 @@
...
+      <<: *aws_connection_info
+    register: vpc_nat_gateway_facts
+    until: vpc_nat_gateway_facts | json_query('result[].tags[].Name') | count == 1
+    retries: 10
+    delay: 5
+  
+  - name: assert tag is added
+    assert:
+      that:
+          - 'vpc_nat_gateway_facts.result.0.tags.Name == "{{ resource_prefix }}-nat"'
```
Comment (tremble) [inline test/integration/targets/ec2_vpc_nat_gateway/tasks/main.yml:95]: It'd be good to add a test for ec2_vpc_nat_gateway_info using filters

```
  - name: Collect nat-gateway facts to check tags
    ec2_vpc_nat_gateway_info:
      filters:
        "tag:Name": "{{ resource_prefix }}-nat"
    register: vpc_nat_gateway_facts

  - name: assert tag is added
    assert:
      that:
          - 'vpc_nat_gateway_facts.result.0.tags.Name == "{{ resource_prefix }}-nat"'
```

Once you've done that it'd be good to add ec2_vpc_nat_gateway_info to  test/integration/targets/ec2_vpc_nat_gateway/aliases

**Review by tremble — CHANGES_REQUESTED** [review]
In addition to the other comments, I ran the full sanity test suite would you be willing to fix a number of small sanity check issues?

```
ERROR: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:471:19: blacklisted-name: Black listed name "_"
ERROR: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:898:15: blacklisted-name: Black listed name "_"
```
(use something like "junk" instead of "_")

```
ERROR: Found 10 validate-modules issue(s) which need to be resolved:
ERROR: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:0:0: doc-default-does-not-match-spec: Argument 'release_eip' in argument_spec defines default as (False) but documentation defines default as (True)
ERROR: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:0:0: doc-default-does-not-match-spec: Argument 'wait_timeout' in argument_spec defines default as (320) but documentation defines default as (300)
ERROR: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:0:0: doc-missing-type: Argument 'state' in argument_spec uses default type ('str') but documentation doesn't define type
ERROR: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:0:0: parameter-type-not-in-doc: Argument 'allocation_id' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:0:0: parameter-type-not-in-doc: Argument 'client_token' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:0:0: parameter-type-not-in-doc: Argument 'eip_address' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:0:0: parameter-type-not-in-doc: Argument 'nat_gateway_id' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:0:0: parameter-type-not-in-doc: Argument 'subnet_id' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:0:0: parameter-type-not-in-doc: Argument 'tags' in argument_spec defines type as 'dict' but documentation doesn't define type
ERROR: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py:0:0: parameter-type-not-in-doc: Argument 'wait_timeout' in argument_spec defines type as 'int' but documentation doesn't define type
```
(I think these all speak for themselves)

Once you've fixed these you can remove the ec2_vpc_nat_gateway exceptions from test/sanity/ignore.txt

File: lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py (hunk 1 above)
Comment (tremble) [reply lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py]: 2.10 :(

**Pull request closed without merging**